scrapy_zyte_api/_poet_item_pipelines.py

```diff
@@ -0,0 +1,88 @@
+from logging import getLogger
+from typing import Any, Set, Type
+
+from itemadapter import ItemAdapter
+from scrapy import Spider
+from scrapy.crawler import Crawler
+from scrapy.exceptions import NotConfigured
+from scrapy.utils.misc import load_object
+from scrapy_poet import InjectionMiddleware
+from web_poet.fields import get_fields_dict
+from web_poet.utils import get_fq_class_name
+from zyte_common_items.fields import is_auto_field
+
+logger = getLogger(__name__)
+
+
+class ScrapyZyteAPIPoetItemPipeline:
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        return cls(crawler)
+
+    def __init__(self, crawler: Crawler):
+        if not crawler.settings.getbool("ZYTE_API_AUTO_FIELD_STATS", False):
+            raise NotConfigured
+
+        raw_url_fields = crawler.settings.getdict("ZYTE_API_AUTO_FIELD_URL_FIELDS", {})
+        self._url_fields = {load_object(k): v for k, v in raw_url_fields.items()}
+        self._seen: Set[Type] = set()
+        self._crawler = crawler
+        self._stats = crawler.stats
+        self._cls_without_url: Set[Type] = set()
+
+    def open_spider(self, spider):
+        for component in self._crawler.engine.downloader.middleware.middlewares:
+            if isinstance(component, InjectionMiddleware):
+                self._registry = component.injector.registry
+                return
+        raise RuntimeError(
+            "Could not find scrapy_poet.InjectionMiddleware among downloader "
+            "middlewares. scrapy-poet may be misconfigured."
+        )
+
+    def process_item(self, item: Any, spider: Spider):
+        cls = item.__class__
+
+        url_field = self._url_fields.get(cls, "url")
+        adapter = ItemAdapter(item)
+        url = adapter.get(url_field, None)
+        if not url:
+            if cls not in self._cls_without_url:
+                self._cls_without_url.add(cls)
+                logger.warning(
+                    f"An item of type {cls} was missing a non-empty URL in "
+                    f"its {url_field!r} field. An item URL is necessary to "
+                    f"determine the page object that was used to generate "
+                    f"that item, and hence print the auto field stats that "
+                    f"you requested by enabling the ZYTE_API_AUTO_FIELD_STATS "
+                    f"setting. If {url_field!r} is the wrong URL field for "
+                    f"that item type, use the ZYTE_API_AUTO_FIELD_URL_FIELDS "
+                    f"setting to set a different field."
+                )
+            return
+
+        page_cls = self._registry.page_cls_for_item(url, cls)
+        cls = page_cls or cls
```

Comment:
I find this line confusing: previously cls was an item class, now it's either a page class or an item class. Do you think it's possible to refactor it, to make the logic a bit more clear? A basic way could be to rename cls to item_cls before this line, and  only introduce `cls` at this line (`cls = page_cls or item_cls` at least makes it clear that cls is now). But maybe there are other approaches, e.g. involving some changes below.

Reply:
https://github.com/scrapy-plugins/scrapy-zyte-api/pull/216/commits/a89f3b72d3e6e28565292c11afed6c8d74c891ff ?